### PR TITLE
refactor/upcoming: [M3-8116] - Event Messages - Part 1

### PR DESCRIPTION
**Pull request by abailly-akamai**
## Description 📝
This PR is part 1 of several contributions to refactor the way we are rendering event messages to our users.

It does not currently affect production. It is behind a feature flag, and once merged meant to sit in develop for some time in order to catch potential formatting issues.

The goals behind the refactor are rather simple:
- Deprecate the formatting & parsing of event messages in favor of composing & rendering JSX messages
- Improve structure, consistency, typing, control and stability of Event Messages 

## Changes  🔄
- Create new utils and JSX components to compose messages
  - Provide coverage and JSDocs
  - Include in Storybook as needed
  - Include future deprecation/deletion comments
- Compose all existing messages and add missing ones
- Implement rendering of new messages behind `eventMessagesV2` feature flag
  - Event table (`/events` + `/linodes/{linodeId}/activity`)
  - Notification Center (popup)
- Improve styling/layout according to approved UX guides
- Update documentation

For more details about changes, see self review and screenshots

## Preview 📷
**Include a screenshot or screen recording of the change**

:bulb: Use `<video src="" />` tag when including recordings in table.

| Before  | After   |
| ------- | ------- |
| Events localhost 3000 (2) | Events localhost |
| Screenshot 2024-05-31 at 10 16 17 | Screenshot 2024-05-31 at 10 15 57 |
| Events localhost 3000 (3) | Events localhost 3000 (4) |

### Prerequisites
- Turn on the `eventMessagesV2` flag

### Verification steps
**Cloud Manager**
- Confirm formatting and displaying of messages in
  - Notification Center (events popup)
  - `/events` page
  - `/linodes/{linodeId}/activity`
 
**Storybook**
- Navigate to http://localhost:6006/?path=/docs/features-event-messages--documentation
  - Verify formatting and typos
- Navigate to http://localhost:6006/?path=/story/components-notifications-snackbar-toast--with-event-message
  - Confirm single message in snackbar

**Docs**
- run `yarn docs`
- navigate to http://localhost:5173/manager/development-guide/15-api-events.html
- confirm page formatting and copy

## As an Author I have considered 🤔

*Check all that apply*

- [x] 👀 Doing a self review
- [ ] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [ ] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a changeset
- [x] 🧪 Providing/Improving test coverage
- [ ] 🔐 Removing all sensitive information from the code and PR description
- [x] 🚩 Using a feature flag to protect the release
- [x] 👣 Providing comprehensive reproduction steps
- [x] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [x] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

### CODE REVIEW

File: packages/manager/src/features/NotificationCenter/NotificationData/RenderEvent.tsx [hunk 1]
```diff
@@ -43,6 +47,33 @@ export const RenderEvent = React.memo((props: RenderEventProps) => {
...
+              <Typography className={unseenEventClass}>
+                {getEventTimestamp(event).toRelative()} | {event.username}
+              </Typography>
+            </Box>
+          </RenderEventStyledBox>
+          <Divider />
+        </>
+      ) : null
+    );
+  }
```
Comment (abailly-akamai) [inline packages/manager/src/features/NotificationCenter/NotificationData/RenderEvent.tsx:75]: New version contains approved UX changes
- User name is not appended to message, and added next to the timestamp, separated by a pipe
- Gravatar size is slightly decreased

File: packages/manager/src/features/NotificationCenter/NotificationSection.tsx [hunk 2]
```diff
@@ -235,7 +235,6 @@ const StyledNotificationItem = styled(Box, {
   shouldForwardProp: omittedProps(['header']),
 })<{ header: string }>(({ theme, ...props }) => ({
   '& p': {
-    color: theme.textColors.headlineStatic,
```
Comment (abailly-akamai) [inline packages/manager/src/features/NotificationCenter/NotificationSection.tsx:238]: removing to make messages the same color as regular copy (not sure why it was overridden)

File: packages/manager/src/features/Events/types.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  [K in EventAction]?: EventMessage;
+};
+
+export type EventMessage = {
+  [S in EventStatus]?: (e: Event) => JSX.Element | string;
+};
+
+export type PartialEventMap<Prefix extends EventActionPrefixes> = {
+  [K in Extract<EventAction, `${Prefix}_${string}`>]: EventMessage;
+};
```
Comment (abailly-akamai) [inline packages/manager/src/features/Events/types.ts:17]: Thanks for this one, @bnussman-akamai - very nifty!

File: docs/development-guide/15-api-events.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  const showToast = (variant: any) =>
+  enqueueSnackbar(failureMessage, {
+    'error',
+  });
+}
+```
+
+Both `action` and `status` are required. The `entity` and `secondary_entity` can optionally be passed to allow for linking. **Note**: it is possible the Event Message formatting will be slightly broken if the action status message expects either but isn't not provided by the instance call.
+
+If a corresponding status does not exist (ex: "failed"), it's quite ok to add it to the Action. Event if not triggered by the API, it can be useful to have a reusable Event Message to use through the App.
```
[the remark this answers is not in the document]
Comment (mjac0bs) [reply docs/development-guide/15-api-events.md:73]: > Even if not triggered by the API, it can be useful to have a reusable Event Message to use through the App.

I'm not totally clear what we mean by this, unless you're referring to triggering the message ourselves (e.g. through toasts)?

File: packages/manager/src/features/NotificationCenter/NotificationData/RenderEvent.tsx [hunk 5]
```diff
@@ -43,6 +47,36 @@ export const RenderEvent = React.memo((props: RenderEventProps) => {
...
+              sx={{ height: 32, minWidth: 32, mt: '3px', width: 32 }}
+              username={event.username}
+            />
+            <Box sx={{ marginTop: '-2px' }}>
+              {messageV2}
+              <Typography
+                className={unseenEventClass}
+                sx={{ fontSize: '0.8rem' }}
+              >
+                {getEventTimestamp(event).toRelative()} | {event.username}
```
Comment (mjac0bs) [inline packages/manager/src/features/NotificationCenter/NotificationData/RenderEvent.tsx:70]: We shouldn't display the divider if there's no username after it.

Screenshot 2024-05-31 at 4 05 56 PM

File: packages/manager/src/features/Events/EventsLanding.tsx [hunk 6]
```diff
@@ -93,13 +104,22 @@ export const EventsLanding = (props: Props) => {
...
+              <Hidden smDown>
+                <TableCell data-qa-events-username-header sx={{ width: 150 }}>
+                  User
+                </TableCell>
+              </Hidden>
+            )}
+            <StyledTableCell sx={{ width: 150 }}>Relative Date</StyledTableCell>
             <Hidden mdDown>
-              <StyledTableCell data-qa-events-time-header>
+              <StyledTableCell data-qa-events-time-header sx={{ width: 150 }}>
```
[the remark this answers is not in the document]
Comment (mjac0bs) [reply packages/manager/src/features/Events/EventsLanding.tsx:122]: I'm all for removing the avatar. 🚀 

User column looks great too, but I did notice that in mobile view, if there is no username, we get a `by  `.

Screenshot 2024-05-31 at 4 13 12 PM

**Review by mjac0bs — COMMENTED** [review]
🥇 For all your work on this, @abailly-akamai. 

I still have some of the factories to review, code to look at more closely, and more testing of some of the edge-case events to do, but from an initial look, this is so nice.

The main thing I noticed as far as bugs is the no username edge case where we're rendering some extra stuff (pipe or "by") that should not be there. Things like linking, styling, responsive layout, and the different places events are present (notification menu vs tables) were looking good in the UI to me.

File: docs/development-guide/15-api-events.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  });
+
+  const showToast = (variant: any) =>
+  enqueueSnackbar(failureMessage, {
+    'error',
+  });
+}
+```
+
+Both `action` and `status` are required. The `entity` and `secondary_entity` can optionally be passed to allow for linking. **Note**: it is possible the Event Message formatting will be slightly broken if the action status message expects either but isn't not provided by the instance call.
```
[the remark this answers is not in the document]
Comment (abailly-akamai) [reply docs/development-guide/15-api-events.md]: 👍 I rewrote this to make it clearer.

**Comment by abailly-akamai** [comment]
Thanks for all the suggestions and the thorough review, @mjac0bs! I know it's a lot to read through! Updated PR with all copy and type suggestions.

**Review by jdamore-linode — APPROVED** [review]
This is such an awesome improvement! Thanks @abailly-akamai.

File: packages/manager/src/features/NotificationCenter/NotificationData/RenderEvent.tsx [hunk 8]
```diff
@@ -43,6 +47,37 @@ export const RenderEvent = React.memo((props: RenderEventProps) => {
...
+              username={event.username}
+            />
+            <Box sx={{ marginTop: '-2px' }}>
+              {messageV2}
+              <Typography
+                className={unseenEventClass}
+                sx={{ fontSize: '0.8rem' }}
+              >
+                {getEventTimestamp(event).toRelative()}
+                {event.username && ` | {event.username}`}
```
[the remark this answers is not in the document]
Comment (abailly-akamai) [reply packages/manager/src/features/NotificationCenter/NotificationData/RenderEvent.tsx]: 🤦 oops - thank you :)

File: packages/manager/cypress/e2e/core/notificationsAndEvents/events.spec.ts [hunk 9]
```diff
@@ -107,6 +112,14 @@ const events: Event[] = eventActions.map((action) => {
...
 
 describe('verify notification types and icons', () => {
+  before(() => {
+    // TODO eventMessagesV2: delete when flag is removed and update test
+    mockAppendFeatureFlags({
+      eventMessagesV2: makeFeatureFlagData(false),
+    });
+    mockGetFeatureFlagClientstream();
+  });
+
```
[the remark this answers is not in the document]
Comment (abailly-akamai) [reply packages/manager/cypress/e2e/core/notificationsAndEvents/events.spec.ts:122]: Great ty!

File: packages/manager/src/features/Events/factories/volume.tsx [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+  volume_delete: {
+    failed: (e) => ``,
+    finished: (e) => ``,
+    notification: (e) => (
+      <>
+        Volume {e.entity?.label} has been <strong>deleted</strong>.
+      </>
+    ),
+    scheduled: (e) => ``,
+    started: (e) => ``,
```
Comment (mjac0bs) [inline packages/manager/src/features/Events/factories/volume.tsx:94]: Do we want to leave these as uncommented empty strings?

(storybook)
image

File: packages/manager/src/features/NotificationCenter/NotificationData/RenderEvent.tsx (hunk 1 above)
Comment (mjac0bs) [reply packages/manager/src/features/NotificationCenter/NotificationData/RenderEvent.tsx:79]: The gravatar size of in-progress events is still a slightly larger size. (Can turn on the MSW to see this, rather than kick off an action that takes a while.)

Screenshot 2024-06-03 at 1 09 53 PM

File: packages/manager/src/features/Events/factories/lke.tsx [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+        The IP ACL for Kubernetes Cluster <EventLink event={e} to="entity" />{' '}
+        has been <strong>disabled</strong>.
+      </>
+    ),
+  },
+  lke_control_plane_acl_update: {
+    notification: (e) => (
+      // `The IP ACL for Kubernetes Cluster${
+      //   e.entity?.label ? ` ${e.entity.label}` : ''
+      // } has been updated.`,
```
Comment (dwiley-akamai) [inline packages/manager/src/features/Events/factories/lke.tsx:67]: Stray comment?

**Review by mjac0bs — APPROVED** [review]
Confirmed:
- Updated Storybook stories load
- Test coverage is added for the changes and tests pass
- The edge case for no usernames with the | or "by " showing up was addressed 

Suggested:
- A lot of minor nits in event message factories. Some typos (that probably existed before this PR), the need for a couple periods, some places where we missed bolding and might have missed linking an entity.
- The need to fix a contrast issue with pre code block styling in dark mode.
- The need to shrink the gravatar size for in-progress events.

Pending the resolution of my 19 or so comments (😅 ), I'm approving this because I've been through all 71 files. Let me know if you want me to circle back to confirm any fixes that are implemented. Thanks, @abailly-akamai!

File: docs/development-guide/15-api-events.md [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+      <>
+        An IP address has been <strong>added</strong> to Linode{' '}
+        <EventLink event={e} to="entity" />.
+      </>
+    ),
+  },
+};
+```
+
+The convention to compose the message is as follow:
```
Comment (dwiley-akamai) [inline docs/development-guide/15-api-events.md:29]:
```suggestion
The convention to compose the message is as follows:
```

**Review by dwiley-akamai — APPROVED** [review]

File: packages/manager/src/features/Events/factories/disk.tsx [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+        <EventLink event={e} to="entity" />.
+      </>
+    ),
+    finished: (e) => (
+      <>
+        Disk <EventLink event={e} to="secondaryEntity" /> has been{' '}
+        <strong>added</strong> to Linode <EventLink event={e} to="entity" />.
+      </>
+    ),
+    notification: () => '',
```
[the remark this answers is not in the document]
Comment (abailly-akamai) [reply packages/manager/src/features/Events/factories/disk.tsx:24]: I am keeping what was there previously. There's frankly quite a few unknowns that will hopefully surface as a result of this epic

[the remark this answers is not in the document]
Comment (abailly-akamai) [reply packages/manager/src/features/Events/factories/disk.tsx:24]: actually you're right, let's fill those - won't hurt

[the remark this answers is not in the document]
Comment (abailly-akamai) [reply packages/manager/src/features/Events/factories/disk.tsx]: done!

File: packages/manager/src/features/Events/factories/disk.tsx [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+    ),
+  },
+  disk_resize: {
+    failed: (e) => (
+      <>
+        A disk on Linode <EventLink event={e} to="entity" /> could{' '}
+        <strong>not</strong> be <strong>resized</strong>.{' '}
+        <Link to="https://www.linode.com/docs/products/compute/compute-instances/guides/disks-and-storage/">
+          Learn more
+        </Link>
```
[the remark this answers is not in the document]
Comment (abailly-akamai) [reply packages/manager/src/features/Events/factories/disk.tsx:145]: Done!

File: packages/manager/src/features/Events/EventRowV2.tsx [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  return (
+    <TableRow data-qa-event-row data-test-id={action}>
+      <TableCell data-qa-event-message-cell parentColumn="Event">
+        {message}
+      </TableCell>
+      <Hidden smDown>
+        <TableCell data-qa-event-username-cell parentColumn="Username">
+          {username ?? 'Unknown'}
```
[the remark this answers is not in the document]
Comment (abailly-akamai) [reply packages/manager/src/features/Events/EventRowV2.tsx]: yup! "Linode" as a fallback is indeed better

File: packages/manager/src/features/Events/EventMessage.tsx [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+          <StyledPre key={i}>{part.slice(1, -1)}</StyledPre>
+        ) : (
+          part
+        )
+      )}
+    </>
+  );
+};
+
+const StyledPre = styled('pre')({
```
[the remark this answers is not in the document]
Comment (abailly-akamai) [reply packages/manager/src/features/Events/EventMessage.tsx]: good catch! done

File: packages/manager/src/features/Events/factories/volume.tsx (hunk 10 above)
Comment (abailly-akamai) [reply packages/manager/src/features/Events/factories/volume.tsx]: added!

File: packages/manager/src/features/Events/factories/lke.tsx (hunk 11 above)
Comment (abailly-akamai) [reply packages/manager/src/features/Events/factories/lke.tsx]: Indeed! removed

**Comment by abailly-akamai** [comment]
Thanks @mjac0bs and @dwiley-akamai - all feedback addressed (either committed from your suggestion or committed myself). Feedback much appreciated!

**Comment by mjac0bs** [comment]
@abailly-akamai There are some hidden comments that github collapsed (which I wish it wouldn't do), that I'm seeing neither marked as Outdated or Resolved. Did those get missed by accident? 

Screenshot 2024-06-03 at 2 33 57 PM

**Comment by abailly-akamai** [comment]
@mjac0bs oops! will certainly address those, thx for catching!

File: packages/manager/src/features/Events/EventMessage.tsx [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+import * as React from 'react';
+
+interface MessageLinkEntity {
+  message: null | string;
+}
+
+/**
+ * Renders a message with inline code blocks.
+ * Meant to be used in the context of an event message.
+ * This component is only used to render {e.message} in the case of potential ticks we want to render as <pre>.
```
[the remark this answers is not in the document]
Comment (abailly-akamai) [reply packages/manager/src/features/Events/EventMessage.tsx:11]: @mjac0bs great point. I think the name is ok but made a ticket to handle this case and make the component a tad more versatile. I want to avoid doing too many string replace since this is in part what got us into this mess but I think "contact support" is a reasonable candidate.

File: packages/manager/src/features/Events/factories/linode.tsx [hunk 18]
```diff
@@ -0,0 +1,540 @@
...
+          </>
+        ) : (
+          ''
+        )}
+        .
+      </>
+    ),
+    notification: (e) => (
+      <>
+        Linode {e.entity?.label} is scheduled to be <strong>cloned</strong>
```
[the remark this answers is not in the document]
Comment (abailly-akamai) [reply packages/manager/src/features/Events/factories/linode.tsx]: done!

File: packages/manager/src/features/Events/factories/linode.tsx [hunk 19]
```diff
@@ -0,0 +1,540 @@
...
+    ),
+  },
+  linode_snapshot: {
+    failed: (e) => (
+      <>
+        Snapshot backup <strong>failed</strong> on Linode{' '}
+        <EventLink event={e} to="entity" />.{' '}
+        <Link to="https://www.linode.com/docs/products/storage/backups/#limits-and-considerations">
+          Learn more about limits and considerations
+        </Link>
```
[the remark this answers is not in the document]
Comment (abailly-akamai) [reply packages/manager/src/features/Events/factories/linode.tsx:511]: done (line 511)
